src/hwdb.rs

```diff
@@ -0,0 +1,97 @@
+use std::ffi::{CString, OsStr};
+use std::io::Result;
+use std::marker::PhantomData;
+use std::os::unix::ffi::OsStrExt;
+
+use libc::c_char;
+
+use ffi;
+use list::List;
+use FromRaw;
+
+/// Rust wrapper for the `udev_hwdb` struct, which provides access to `udev`'s
+/// hardware database API.
+///
+/// Like the `udev` struct, `udev_hwdb` is refcounted and automatically managed
+/// by the Rust wrapper.
+pub struct Hwdb {
+    hwdb: *mut ffi::udev_hwdb,
+}
+
+impl Clone for Hwdb {
+    fn clone(&self) -> Self {
+        unsafe { Self::from_raw(ffi::udev_hwdb_ref(self.hwdb)) }
+    }
+}
+
+impl Drop for Hwdb {
+    fn drop(&mut self) {
+        unsafe { ffi::udev_hwdb_unref(self.hwdb) };
+    }
+}
+
+as_ffi!(Hwdb, hwdb, ffi::udev_hwdb, ffi::udev_hwdb_ref);
+
+impl Hwdb {
+    /// Creates a new Hwdb context.
+    pub fn new() -> Result<Self> {
```

Comment:
N.B.: The fact that `udev_hwdb_new` doesn't technically require a `udev` handle is probably an implementation detail. I can rewrite this to take a `&UDev` that it could then make a ref-copy of, if you'd like.

Reply:
Looks like I got too eager: old-ish versions of `udev` check the pointer, even though they don't use it: https://github.com/systemd/systemd/blob/d9b02e16971e13c36311b5ad3cb4c040a5024bf9/src/libudev/libudev-hwdb.c#L43

Reply:
Changed this to use a copy of a `udev` pointer. That pointer can become stale if its owning `Udev` gets dropped, but that shouldn't matter since even the older versions that check the pointer don't seem to use it.